ui: `go generate` fails on Linux because of osx-specific dependency

While working on #4548 I needed to run `go generate` and encountered the following:

```
npm install --progress=false
npm WARN deprecated lodash@1.0.2: lodash@<3.0.0 is no longer maintained. Upgrade to lodash@^4.0.0.
npm WARN deprecated object-pick@0.1.2: Deprecated as of version 0.1.2 please use object.pick instead.
npm WARN engine boom@2.10.1: wanted: {"node":">=0.10.40"} (current: {"node":"0.10.25","npm":"1.4.21"})
npm WARN engine cryptiles@2.0.5: wanted: {"node":">=0.10.40"} (current: {"node":"0.10.25","npm":"1.4.21"})
npm WARN engine deep-extend@0.4.1: wanted: {"node":">=0.12.0","iojs":">=1.0.0"} (current: {"node":"0.10.25","npm":"1.4.21"})
npm WARN engine hawk@3.1.3: wanted: {"node":">=0.10.32"} (current: {"node":"0.10.25","npm":"1.4.21"})
npm WARN engine hoek@2.16.3: wanted: {"node":">=0.10.40"} (current: {"node":"0.10.25","npm":"1.4.21"})
npm ERR! notsup Unsupported
npm ERR! notsup Not compatible with your operating system or architecture: fsevents@1.0.7
npm ERR! notsup Valid OS:    darwin
npm ERR! notsup Valid Arch:  any
npm ERR! notsup Actual OS:   linux
npm ERR! notsup Actual Arch: x64

npm ERR! System Linux 4.2.0-27-generic
npm ERR! command "/usr/bin/nodejs" "/usr/bin/npm" "install" "--progress=false"
npm ERR! cwd /home/kena/go/src/github.com/cockroachdb/cockroach/ui
npm ERR! node -v v0.10.25
npm ERR! npm -v 1.4.21
npm ERR! code EBADPLATFORM
npm WARN deprecated graceful-fs@3.0.8: graceful-fs version 3 and before will fail on newer node releases. Please update to graceful-fs@^4.0.0
 as soon as possible.
npm WARN deprecated graceful-fs@1.2.3: graceful-fs version 3 and before will fail on newer node releases. Please update to graceful-fs@^4.0.0
 as soon as possible.
npm ERR! 
npm ERR! Additional logging details can be found in:
npm ERR!     /home/kena/go/src/github.com/cockroachdb/cockroach/ui/npm-debug.log
npm ERR! not ok code 0
Makefile:102: recipe for target 'npm.installed' failed
make: *** [npm.installed] Error 1
ui/ui.go:21: running "make": exit status 2
```

That's odd. How come it works in CircleCI?

On Mon, Feb 22, 2016 at 10:49 AM, kena <notifications@github.com> wrote:

> While working on #4548
> <https://github.com/cockroachdb/cockroach/pull/4548> I needed to run go
> generate and encountered the following:
>
> npm install --progress=false
> npm WARN deprecated lodash@1.0.2: lodash@<3.0.0 is no longer maintained. Upgrade to lodash@^4.0.0.
> npm WARN deprecated object-pick@0.1.2: Deprecated as of version 0.1.2 please use object.pick instead.
> npm WARN engine boom@2.10.1: wanted: {"node":">=0.10.40"} (current: {"node":"0.10.25","npm":"1.4.21"})
> npm WARN engine cryptiles@2.0.5: wanted: {"node":">=0.10.40"} (current: {"node":"0.10.25","npm":"1.4.21"})
> npm WARN engine deep-extend@0.4.1: wanted: {"node":">=0.12.0","iojs":">=1.0.0"} (current: {"node":"0.10.25","npm":"1.4.21"})
> npm WARN engine hawk@3.1.3: wanted: {"node":">=0.10.32"} (current: {"node":"0.10.25","npm":"1.4.21"})
> npm WARN engine hoek@2.16.3: wanted: {"node":">=0.10.40"} (current: {"node":"0.10.25","npm":"1.4.21"})
> npm ERR! notsup Unsupported
> npm ERR! notsup Not compatible with your operating system or architecture: fsevents@1.0.7
> npm ERR! notsup Valid OS:    darwin
> npm ERR! notsup Valid Arch:  any
> npm ERR! notsup Actual OS:   linux
> npm ERR! notsup Actual Arch: x64
>
> npm ERR! System Linux 4.2.0-27-generic
> npm ERR! command "/usr/bin/nodejs" "/usr/bin/npm" "install" "--progress=false"
> npm ERR! cwd /home/kena/go/src/github.com/cockroachdb/cockroach/ui
> npm ERR! node -v v0.10.25
> npm ERR! npm -v 1.4.21
> npm ERR! code EBADPLATFORM
> npm WARN deprecated graceful-fs@3.0.8: graceful-fs version 3 and before will fail on newer node releases. Please update to graceful-fs@^4.0.0
>  as soon as possible.
> npm WARN deprecated graceful-fs@1.2.3: graceful-fs version 3 and before will fail on newer node releases. Please update to graceful-fs@^4.0.0
>  as soon as possible.
> npm ERR!
> npm ERR! Additional logging details can be found in:
> npm ERR!     /home/kena/go/src/github.com/cockroachdb/cockroach/ui/npm-debug.log
> npm ERR! not ok code 0
> Makefile:102: recipe for target 'npm.installed' failed
> make: *** [npm.installed] Error 1
> ui/ui.go:21: running "make": exit status 2
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/issues/4555>.
>

After upgrading to a more recent version of `master` I run into a different issue:
```
npm install --progress=false
/usr/bin/env: node: No such file or directory
Makefile:102: recipe for target 'npm.installed' failed
```

Now I can work this issue by symlinking `/usr/bin/nodejs` to `~/bin/node` but again there's too much magic.

I do not know why it works in CI but it definitely does not work out of the box on this plain, no-frills, out-of-the-box ubuntu 15.

I suggest you run it inside the builder container.

On Mon, Feb 22, 2016 at 10:54 AM, kena <notifications@github.com> wrote:

> After upgrading to a more recent version of master I run into a different
> issue:
>
> npm install --progress=false
> /usr/bin/env: node: No such file or directory
> Makefile:102: recipe for target 'npm.installed' failed
>
> Now I can work this issue by symlinking /usr/bin/nodejs to ~/bin/node but
> again there's too much magic.
>
> I do not know why it works in CI but it definitely does not work out of
> the box on this plain, no-frills, out-of-the-box ubuntu 15.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/cockroachdb/cockroach/issues/4555#issuecomment-187241466>
> .
>

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -27,19 +27,27 @@
 
 Note that if you edit a `.proto` or `.ts` file, you will need to
 manually regenerate the associated `.pb.{go,cc,h}` or `.js` files
-using `go generate ./...`.  `go generate` requires a collection of
-node modules which are installed via npm. If you don't have npm, it
-typically comes with node. To get it via homebrew: `brew install node`
-If you're not using homebrew, make sure you install both
-[node.js](https://nodejs.org/) and [npm](https://www.npmjs.com/).  If
-you plan on working on the UI, check out [the ui readme](ui).
+using `go generate ./...`.
+
+We advise to run `go generate` using our embedded Docker
+setup. `build/builder.sh` is a wrapper script designed to make this
+convenient. You can run `build/builder.sh go generate ./...` from the
+repository root to get the intended result.
+
+If you want to run it outside of Docker, `go generate` requires a
+collection of Node.js modules which are installed via npm.
+
+If you plan on working on the UI, check out [the ui readme](ui).
 
 To add or update a go dependency:
 - `(cd $GOPATH/src && go get -u ./...)` to update the dependencies or
   `go get {package}` to add a dependency
 - `glock save github.com/cockroachdb/cockroach` to update the
   GLOCKFILE
-- `go generate ./...` to update generated files
+- `go generate ./...` to update generated files -- prefer `go generate
+  ./the-updated-package` instead of `...` when possible to avoid
+  re-generating files in directories where you haven't made any
+  changes.
 - create a PR with all the changes
 
 ### Style guide

--- a/CONTRIBUTING.md
+++ b/CONTRIBUTING.md
@@ -3,33 +3,55 @@
 ### Getting and building
 
 1.  Install the following prerequisites, as necessary:
-    - A C++ compiler that supports C++11 (GCC 4.9+ and clang 3.6+ are known to work). On Mac OS X, Xcode should suffice.
-    - A [Go environment](http://golang.org/doc/code.html) with a 64-bit version of Go 1.6. You can download the [Go binary](https://golang.org/dl/) directly from the official site. On OS X, you can also use [homebrew](http://brew.sh): `brew install go`. Be sure to set the `$GOPATH` and `$PATH` environment variables as described [here](https://golang.org/doc/code.html#GOPATH).
-    - Git 1.8+
+	- A C++ compiler that supports C++11 (GCC 4.9+ and clang 3.6+ are
+	  known to work). On Mac OS X, Xcode should suffice.
+	- A [Go environment](http://golang.org/doc/code.html) with a
+	  64-bit version of Go 1.6. You can download the
+	  [Go binary](https://golang.org/dl/) directly from the official
+	  site. On OS X, you can also use [homebrew](http://brew.sh):
+	  `brew install go`. Be sure to set the `$GOPATH` and `$PATH`
+	  environment variables as described
+	  [here](https://golang.org/doc/code.html#GOPATH).
+	- Git 1.8+
 
 2.  Get the CockroachDB code:
 
-    ```bash
-    go get -d github.com/cockroachdb/cockroach
-    cd $GOPATH/src/github.com/cockroachdb/cockroach
-    ```
+	```bash
+	go get -d github.com/cockroachdb/cockroach
+	cd $GOPATH/src/github.com/cockroachdb/cockroach
+	```
 
-3.  Run `make build`, `make test`, or anything else our Makefile offers. Note that the first time you run `make`, it can take some time to download and install various dependencies.
+3.  Run `make build`, `make test`, or anything else our Makefile
+	offers. Note that the first time you run `make`, it can take some
+	time to download and install various dependencies.
 
-Note that if you edit a `.proto` or `.ts` file, you will need to manually regenerate the associated `.pb.{go,cc,h}` or `.js` files using `go generate ./...`.
-`go generate` requires a collection of node modules which are installed via npm. If you don't have npm, it typically comes with node. To get it via homebrew:
-`brew install node`
-If you're not using homebrew, make sure you install both [node.js](https://nodejs.org/) and [npm](https://www.npmjs.com/).
-If you plan on working on the UI, check out [the ui readme](ui).
+Note that if you edit a `.proto` or `.ts` file, you will need to
+manually regenerate the associated `.pb.{go,cc,h}` or `.js` files
+using `go generate ./...`.  `go generate` requires a collection of
+node modules which are installed via npm. If you don't have npm, it
+typically comes with node. To get it via homebrew: `brew install node`
+If you're not using homebrew, make sure you install both
+[node.js](https://nodejs.org/) and [npm](https://www.npmjs.com/).  If
+you plan on working on the UI, check out [the ui readme](ui).
 
 To add or update a go dependency:
-- `(cd $GOPATH/src && go get -u ./...)` to update the dependencies or `go get {package}` to add a dependency
-- `glock save github.com/cockroachdb/cockroach` to update the GLOCKFILE
+- `(cd $GOPATH/src && go get -u ./...)` to update the dependencies or
+  `go get {package}` to add a dependency
+- `glock save github.com/cockroachdb/cockroach` to update the
+  GLOCKFILE
 - `go generate ./...` to update generated files
 - create a PR with all the changes
 
 ### Style guide
-We're following the [Google Go Code Review](https://code.google.com/p/go-wiki/wiki/CodeReviewComments) fairly closely. In particular, you want to watch out for proper punctuation and capitalization in comments. We use two-space indents in non-Go code (in Go, we follow `gofmt` which indents with tabs). Format your code assuming it will be read in a window 100 columns wide. Wrap code and comments at 100 characters unless doing so makes the code less legible.
+
+We're following the
+[Google Go Code Review](https://code.google.com/p/go-wiki/wiki/CodeReviewComments)
+fairly closely. In particular, you want to watch out for proper
+punctuation and capitalization in comments. We use two-space indents
+in non-Go code (in Go, we follow `gofmt` which indents with
+tabs). Format your code assuming it will be read in a window 100
+columns wide. Wrap code and comments at 100 characters unless doing so
+makes the code less legible.
 
 ### Code review workflow
 
@@ -43,8 +65,11 @@ We're following the [Google Go Code Review](https://code.google.com/p/go-wiki/wi
 
   `git checkout -b update-readme`
 
-+ Hack away and commit your changes locally using `git add` and `git commit`. Remember to write tests! The following are helpful for running specific subsets of tests:
-  ```
++ Hack away and commit your changes locally using `git add` and `git
+  commit`. Remember to write tests! The following are helpful for
+  running specific subsets of tests:
+
+  ```bash
   make test
   # Run all tests in ./storage
   make test PKG=./storage
@@ -52,8 +77,8 @@ We're following the [Google Go Code Review](https://code.google.com/p/go-wiki/wi
   make test PKG=./kv TESTS='^TestFoo' TESTTIMEOUT=10s
   ```
 
-  When you're ready to commit, do just that with a succinct title and informative
-  message. For example,
+  When you're ready to commit, do just that with a succinct title and
+  informative message. For example,
 
   ```bash
   $ git commit
@@ -65,37 +90,70 @@ We're following the [Google Go Code Review](https://code.google.com/p/go-wiki/wi
   > Fixed some formatting.'
   ```
 
-+ Run the whole CI test suite locally: `./build/circle-local.sh`. This requires the Docker setup; if you don't have/want that, `go generate ./... && make check test testrace` is a good first approximation.
++ Run the whole CI test suite locally: `./build/circle-local.sh`. This
+  requires the Docker setup; if you don't have/want that, `go generate
+  ./... && make check test testrace` is a good first approximation.
 
-+ When you’re ready for review, groom your work: each commit should pass tests and contain a substantial (but not overwhelming) unit of work. You may also want to `git fetch origin` and run `git rebase -i --exec "make check test" origin/master` to make sure you're submitting your changes on top of the newest version of our code. Next, push to your fork:
++ When you’re ready for review, groom your work: each commit should
+  pass tests and contain a substantial (but not overwhelming) unit of
+  work. You may also want to `git fetch origin` and run `git rebase -i
+  --exec "make check test" origin/master` to make sure you're
+  submitting your changes on top of the newest version of our
+  code. Next, push to your fork:
 
   `git push -u <yourfork> update-readme`
 
-+ Then [create a pull request using GitHub’s UI](https://help.github.com/articles/creating-a-pull-request).
++ Then
+  [create a pull request using GitHub’s UI](https://help.github.com/articles/creating-a-pull-request).
 
-+ If you get a test failure in CircleCI, check the Test Failure tab to see why the test failed. When the failure is logged in `excerpt.txt`, you can find the file from the Artifacts tab and see log messages. (You need to sign in to see the Artifacts tab.)
++ If you get a test failure in CircleCI, check the Test Failure tab to
+  see why the test failed. When the failure is logged in
+  `excerpt.txt`, you can find the file from the Artifacts tab and see
+  log messages. (You need to sign in to see the Artifacts tab.)
 
-+ Address feedback in new commits. Wait (or ask) for new feedback on those commits if they are not straightforward. An `LGTM` ("looks good to me") by someone qualified is usually posted when you're free to go ahead and merge. Most new contributors aren't allowed to merge themselves; in that case, we'll do it for you. You may also be asked to re-groom your commits.
++ Address feedback in new commits. Wait (or ask) for new feedback on
+  those commits if they are not straightforward. An `LGTM` ("looks
+  good to me") by someone qualified is usually posted when you're free
+  to go ahead and merge. Most new contributors aren't allowed to merge
+  themselves; in that case, we'll do it for you. You may also be asked
+  to re-groom your commits.
 
 
 ### Debugging
 
 Peeking into a running cluster can be done in several ways:
 
-* the [net/trace](https://godoc.org/golang.org/x/net/trace) endpoint at `/debug/requests`.
-  It has a breakdown of the recent traced requests, in particularly slow ones. Two families are traced: `node` and `coord`, the former (and likely more interesting one) containing what happens inside of `Node`/`Store`/`Replica` and the other inside of the coordinator (`TxnCoordSender`).
-* [pprof](https://golang.org/pkg/net/http/pprof/) gives us (among other things) heap and cpu profiles; [this golang blog post](http://blog.golang.org/profiling-go-programs) explains it extremely well and [this one by Dmitry Vuykov](https://software.intel.com/en-us/blogs/2014/05/10/debugging-performance-issues-in-go-programs) goes into even more detail. Two caveats: the `cockroach` binary passed to `pprof` must be the same as the one creating the profile (not true on OSX in acceptance tests!), and the HTTP client used by `pprof` doesn't simply swallow self-signed certs (relevant when using SSL). For the latter, a workaround of the form
+* the [net/trace](https://godoc.org/golang.org/x/net/trace) endpoint
+  at `/debug/requests`.  It has a breakdown of the recent traced
+  requests, in particularly slow ones. Two families are traced: `node`
+  and `coord`, the former (and likely more interesting one) containing
+  what happens inside of `Node`/`Store`/`Replica` and the other inside
+  of the coordinator (`TxnCoordSender`).
+* [pprof](https://golang.org/pkg/net/http/pprof/) gives us (among
+  other things) heap and cpu profiles;
+  [this golang blog post](http://blog.golang.org/profiling-go-programs)
+  explains it extremely well and
+  [this one by Dmitry Vuykov](https://software.intel.com/en-us/blogs/2014/05/10/debugging-performance-issues-in-go-programs)
+  goes into even more detail. Two caveats: the `cockroach` binary
+  passed to `pprof` must be the same as the one creating the profile
+  (not true on OSX in acceptance tests!), and the HTTP client used by
+  `pprof` doesn't simply swallow self-signed certs (relevant when
+  using SSL). For the latter, a workaround of the form
 
   ```
   go tool pprof cockroach <(curl -k https://$(hostname):26257/debug/pprof/profile)
   ```
+
   will do the trick.
 
-An easy way to locally run a workload against a cluster are the acceptance tests.
-For example,
+An easy way to locally run a workload against a cluster are the
+acceptance tests.  For example,
 
 ```bash
 make acceptance TESTS='TestPut$$' TESTFLAGS='-v -d 1200s -l .' TESTTIMEOUT=1210s
 ```
 
-runs the `Put` acceptance test for 20 minutes with logging (useful to look at the stacktrace in case of a node dying). When it starts, all the relevant commands for `pprof`, `trace` and logs are logged to allow for convenient inspection of the cluster.
+runs the `Put` acceptance test for 20 minutes with logging (useful to
+look at the stacktrace in case of a node dying). When it starts, all
+the relevant commands for `pprof`, `trace` and logs are logged to
+allow for convenient inspection of the cluster.

